server/index.js

```diff
@@ -135,14 +129,121 @@ var infer = function(req, res) {
 		var conf = req.model.getConfiguration();
 		if(conf.expiration) ret.expiration = conf.expiration;
 
-        res.json(ret);
+        cb(null, ret);
     }).catch(function(e) {
-        res.json({
-            error: e
-        });
+        cb(e, null)
     }).finally(function() {
         roboflow.tf.dispose(tensor);
     });
+}
+
+var infer = function(req, res) {
+	if(req.query.format && req.query.format == "image") {
+		return res.json({
+			error: "The inference server does not yet support returning a visualization of the prediction."
+		});
+	}
+
+    roboflow.tf.tidy(() => {
+
+        var start = Date.now();
+
+        var buffer = Buffer.from(req.body, "base64");
+        var arr = new Uint8Array(buffer);
+
+        var tensor;
+        try {
+            // works for decoding 3 channel (rgb) and 4 channel (rgba) images
+            tensor = roboflow.tf.node.decodeImage(arr, 3);
+        } catch(e) {
+            // handle 1 channel (grayscale) images
+            // which throw an error if you try to force them to be decoded with 3 channels
+            var channel = roboflow.tf.node.decodeImage(arr, 1);
+            tensor = roboflow.tf.stack([
+                channel,
+                channel,
+                channel
+            ], 2).squeeze();
+            roboflow.tf.dispose(channel);
+        }
+
+        var configuration = {
+            max_objects: Number.MAX_SAFE_INTEGER,
+            overlap: 0.3,
+            nms_threshold: 0.3,
+            threshold: 0.4,
+            tile: false,
+            tile_rows: 1,
+            tile_cols: 1,
+        };
+
+        if(req.query.overlap) {
+            req.query.overlap = parseFloat(req.query.overlap);
+            if(req.query.overlap > 1) req.query.overlap /= 100;
+            configuration.overlap = req.query.overlap;
+            configuration.nms_threshold = req.query.overlap;
+        }
+
+        if(req.query.confidence) {
+            req.query.confidence = parseFloat(req.query.confidence);
+            if(req.query.confidence > 1) req.query.confidence /= 100;
+            configuration.threshold = req.query.confidence;
+        }
+
+        //takes a tile query parameter like so tile=3x5
+        if(req.query.tile) {
+            if(false){
+                return res.json({
+                    error: "The inference server does not support this query parameter without a hosted image."
+                });
+            }
+            configuration.tile = true;
+
+            //ensure the tile query param looks like tile=300
+            if (!req.query.tile.match(/([0-9]+)/)){
+                return res.json({
+                    error: "Tile query parameter improperly formatted."
```

Comment:
This error message could be more helpful; we should tell the user what the format should be.

Would it be hard to take non-square tiles if the user wanted to maintain aspect ratio? (Eg if they trained their model on non-square tiles which got squished in the resize step to square we'd want that same thing to happen during inference for most accurate results.)

Recommendation would be also accepting eg `tile=300x200` format (and if they don't have the `x###` part we treat it as a square and use the first value for both dimensions.

Reply:
Got the errors handled better.

Possible but don't know if that affects our performance. Will punt to a v2.

Also punting rectangular tiling to a v2 since the current tensor logic doesn't slice up tiles correctly when given rectangular dimensions.